Tenants Feature Branch

johnswanson opened the pull request:
This is a less-than-MVP version of Tenants. We want something that we can merge to `master`, hidden behind the `tenants` feature, that will not affect existing customers, but allow us to quickly iterate towards something release-able. So we're looking for something internally complete, `master`-quality, but not necessarily ready to throw out the door.

## User Stories

### As an administrator, I can create new Tenants to define exclusive groups of external users.

These tenants have a name and a slug. There's a minimal UI for managing these tenants.

### As an administrator, I can create a new Tenant User to allow someone from this tenant to log in.

There is a minimal UI for managing these users. It is not currently possible to change an existing user into a tenant user, vice versa, or transfer a user from one tenant to another.

### As an administrator, I can modify permissions on a new group, the "All External Users" group, to control what these users can see.

Note that right now, all collections are locked down - it is prohibited for a group of "external users" to get *any* access to a collection in the system. (Later, we will release "shared tenant collections", allowing external users to receive access to these special collections.)

You can however give this group any data permissions you choose.

### As a tenant user, I can log in and use the application according to these permissions.

Due to the lack of collection permissions, this must currently be tested using ad-hoc queries.

### A new user can be assigned to a tenant via JWT

Once a user is created, tenant cannot be re-assigned and JWT auth will fail if it provides a different tenant than is actually assigned.

### As an administrator, I can use the `@tenant.slug` attribute for database routing, sandboxing, or impersonation

This is a special, pre-assigned attribute that is always and only available when multi-tenancy is enabled. It cannot be overridden, e.g. by user attributes.

## Not included:

- tenant collections (any kind)
- audit events
- tenant custom attributes

johnswanson committed:
Tenants, External Users, and tenant based JWT with Audit Events

Allow the creation of tenant users

While working on this, I realized that my approach to the "All External
Users" group and renaming the "All Users" group was untenable.

I had been planning to just update these things in the database (create
AEU, rename AU to AIU) in the setter function for the `:use-tenants`
setting. The issue here is caching: we don't want to need to actually
reach out to the database every time we need to fetch the All Users or
All External Users, but it would also be quite unfortunate if one
Metabase instance in a cluster had an outdated view of, e.g. the ID of
"All External Users".

(Although - I'm not really sure how much we use the *name* of "All
Users" when we fetch `(all-users)`, so that might be worth a look.)

For now, I've modified things to just update things dynamically in the
`GET /api/permissions/group` and `GET /api/permissions/group/:id`
endpoints: we show/hide the "All External Users" group, and rename the
"All Users" group, based on whether Tenants are enabled.

Allow the creation of tenants

Listing tenants / tenant members will be a followup PR

lint

doh missing require

fix tests

add a test for enabling tenants changing groups

fix tests, slight change to tenant creation API

kick ci

fix another test

give minimum permissions to ext groups

fix more tests

impl user stategy setting ui

more API endpoints

add more tests, fix a very silly migration

slug validation

fix mariadb/mysql

lint migrations file

kondo

fix test

replace tenant members API with `/user?tenant_id=`

also realized that the `/user/recipients` endpoint needed to be modified
for tenants: tenant users should not see recipients that are not part of
the same tenant. (reminder for later: we will need to add safeguards
around actually creating those pulses, not just assume the API won't
send bad recipients!!)

IF you are an admin, `/api/user` respects a `tenant_id` query param and returns:

- "normal" (non-tenant) users if it's `nil`

- specific-tenant users if not

If you are *not* an admin, `/api/user` ignores the `tenant_id` passed
and only returns users in the same tenant.

`/api/user/recipients` will:

- return ALL users if you are a superuser (both tenant and non-tenant)

- return only users in the SAME tenant as you if not

doh

tenant crud wip

integrate endpoints, fix type errors

Allow getting all external users

previously you could do only

```
GET /api/user                => internal users
GET /api/user?tenant_id=null => internal users
GET /api/user?tenant_id=123  => users in tenant 123
```

now you can do

```
GET /api/user                  => internal users
GET /api/user?tenant_id=null   => internal users
GET /api/user?tenant_id=123    => users in tenant 123
GET /api/user?tenancy=all      => internal + external users
GET /api/user?tenancy=internal => internal users
GET /api/user?tenancy=external => external users (in all tenants)
```

If you specify both `tenancy` and `tenant_id` (regardless of the values
passed) you'll get a 400.

fix my bad rebase

fix bad rebase more

remove #p

mark tenants as `is_active` or not

initial external users management

remove permissions check on the trash endpoint

I don't know why this was added in the first place (I mean, I think I
added it, but I don't know why I did).

The trash is always accessible to everyone. We gate the visibility of
things *in* the trash, but the trash itself is always readable.

no personal collection for tenant users

remove tenant query inside oss

tenant user routes

make it so that you cannot remove users from the external users group

do not send email to tenant users

no login for inactive tenants

disable collection permissions, don't show sent email modal on user success with SMTP

remove all users group has higher permissions warnings for external users group

moving is external checks behind plugins

unit test cleanup, add simple test

read only application permissions for external users

set tenant on login with JWT

incredibly hacky tenant attribute

separate `:attributes` from `:login_attributes`

thinking here is that `login_attributes` are the things you set on the
user, where `attributes` are all the attributes of the user including
things like tenant slug or maybe someday locale etc.

fix build

fix token check

create a tenant collection in a migration

this works, but the tenant collection is always visible to admins, even
if the setting is off.

hide tenant collection from normal users

fixing build and FE rebase

Delete stray "howdy"

Lol, sorry about that.

fix tests

throw instead of silent failure on no tenant

superuser can't be tenant

fix up tenant collections

tenant collection now includes subcollections.

Tenant collections are hidden if the feature isn't enabled.

tenant collections marked in API

bugfix and current user can write any coll

don't add `attributes` key to all users

just the current user on login

invalidate groups when changing user strategy

read only user

{{ tenant_slug }} user attribute

fix be test

rip out tenant collection stuff on BE

fix BE tests

lint

fix magic string

no tenantifying existing users for now

also clean up a complicated, confusing fn.

starting e2e

fixing e2e, adjusting @tenant user attribute

fix permissions editor, add tenant.name e2e test

allow more than one user per tenant

:facepalm:

allow getting only active tenants

*actually* don't replace existing values with nil

getting e2e ready for tenant user provisioning

tenants management ux

e2e coverage on sandboxing, deactivating tenants

replace {{tenant_slug}} with `@tenant.slug`

filter out other-tenant users from pulse creation

rename any existing `@.+` user attributes

prohibit new user attributes starting with `@`

fix migration

ugh, forgot to actually commit this file

add more tests

kondo

use the `tenants` feature

fix a test - missed one `:advanced-permissions`

e2e test coverage, updating tenant attribute to @tenant.slug

Partial Type Fixes

types green

unit tests

adding groups to e2e test, maybe fixing tenant feature check?

Add tenants to tracked snowplow features

Fixing Admin ui

Modernizing tenants page

separate external group in permissions sidebar

permissions editor updates

hardcode that *everyone* has trash access

shame... bad imports

adjusting Main Navbar. using globe icon

CR response

fix tests

fix attributes when tenants are disabled

lock up the 'all external users' collection

tenants route guard

add tests for missing feature and oss

fixing some unit tests, and type errors

set token features in default.cy.snap

unit and e2e test adjustment

another pass on e2e

fix e2e and build

fix a bad merge

Green CI?

reset token after restoring DB

move a test to fix the test

kondo

Move `/api/ee/tenants` to `/api/ee/tenant`

PR Feedback, part 1

move route functions into ee

ADM-1105 permissions editor adjustments for external users

ADM 1104 do not show invite button on external users when there is no tenant

adjusting unit tests

fix no results message for deactivated users page

center no results message

Fix migrations IDs / related tests

adjusting no results message for tehants

fixing unit tests, adding e2e test for permissions

chore: add timestamp columns to tenants (#60575)

* chore: add timestamp columns to tenants

* fixup! add hooks

* kick ci

Auto provision tenants, respect tenants feature (#60571)

deactivate tenant users along with their tenants (#60552)

* deactivate tenant users along with their tenants

Tenant users from deactivated tenants were already prevented from
logging in (we checked the activation status of their tenant) but for
bookkeeping it's probably best to actually deactivate the users
themselves along with their tenants, and reactivate them if the tenant
is reactivated.

* invalidate users when updating a tenant, adjust external users list screen

* e2e test

---------

Co-authored-by: Nick Fitzpatrick <nickfitz.582@gmail.com>

audit events for tenants (#60647)

npfitz commented on e2e/snapshot-creators/default.cy.snap.js:
```diff
@@ -361,6 +361,9 @@ describe("snapshots", () => {
 });
 
 function getDefaultInstanceData() {
+  cy.request("PUT", "/api/setting", {
```
We do this to ensure that the "All External Users" collection is returned from "/api/permissions/groups"

npfitz commented on enterprise/frontend/src/metabase-enterprise/sandboxes/components/LoginAttributesWidget.tsx:
```diff
@@ -24,9 +28,14 @@ export const LoginAttributesWidget = ({
   return (
     <FormField className={className} style={style} title={title}>
       <MappingEditor
+        disabledValues={getDisabledTenantUserAttribute(data)}
```
I don't really love how this was done, but basically we want to show the `@tenant.slug` attribute if it's applicable. Instead of passing props in, I'm taking advantage of the fact that there is a form context I can look at and build the information that way.

npfitz commented on enterprise/frontend/src/metabase-enterprise/sandboxes/utils.tsx:
```diff
@@ -0,0 +1,65 @@
+  </Group>
+);
+
+export const getDisabledTenantUserAttribute = (
```
This is for displaying a disabled row in the mapping editor.

npfitz commented on enterprise/frontend/src/metabase-enterprise/tenants/containers/TenantActivationModal.tsx:
```diff
@@ -0,0 +1,62 @@
+    <ConfirmModal
+      opened
+      title={t`Reactivate ${tenant.name}?`}
+      message="TODO: SOME KIND OF COPY HERE"
```
Whoops

npfitz commented on frontend/src/metabase/admin/people/components/GroupDetail.tsx:
```diff
@@ -146,6 +152,19 @@ export const GroupDetail = ({
 };
 
 const GroupDescription = ({ group }: { group: Group }) => {
+  if (PLUGIN_TENANTS.isExternalUsersGroup(group)) {
```
Should cover this in a test

npfitz commented on frontend/src/metabase/api/user.ts:
```diff
@@ -51,7 +51,8 @@ export const userApi = Api.injectEndpoints({
       }),
-      invalidatesTags: (_, error) => invalidateTags(error, [listTag("user")]),
+      invalidatesTags: (_, error) =>
+        invalidateTags(error, [listTag("user"), listTag("tenant")]),
```
Gotta fix this :/

npfitz commented on frontend/src/metabase/components/NewItemMenu/NewItemMenu.tsx:
```diff
@@ -85,14 +88,16 @@ const NewItemMenu = ({
-        dispatch(setOpenModal("dashboard"));
-      },
-    });
+    if (canWriteToCollections) {
```
Should throw a unit test on here

npfitz commented on frontend/src/metabase/nav/containers/MainNavbar/MainNavbarContainer/MainNavbarView.tsx:
```diff
@@ -223,30 +224,32 @@ export function MainNavbarView({
-              <CollectionSectionHeading
-                handleCreateNewCollection={handleCreateNewCollection}
-              />
+          {regularCollections.length > 0 && (
```
See if we can cover this with a unit test

johnswanson commented on enterprise/backend/src/metabase_enterprise/sso/integrations/jwt.clj:
```diff
@@ -31,7 +32,11 @@
               :sso_source       :jwt
-              :login_attributes user-attributes}]
+              :login_attributes user-attributes
+              :tenant_id        (when tenant-slug
```
I don't _think_ we want to prevent it. Like I think you still want to be able to create regular users via JWT. I guess maybe we could make it an option sometime? Enable JWT user provisioning but only for tenant users, block it for regular users?

johnswanson commented on enterprise/backend/src/metabase_enterprise/sandbox/models/params/field_values.clj:
```diff
@@ -54,7 +54,7 @@
   [{:keys [table_id] :as _field}]
   (when-let [gtap (table-id->gtap table_id)]
-    (let [login-attributes     (:login_attributes @api/*current-user*)
+    (let [login-attributes     (:attributes @api/*current-user*)
```
Nice, I like it - will do.

iethree reviewed (approved):
Frontend looks really really good. Very nice work.

Lots of suggestions/nits/questions. Nothing blocking.

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/containers/EditTenantModal/EditTenantModal.tsx:
I think we want to move away from separate `containers/` directories

iethree commented on enterprise/frontend/src/metabase-enterprise/api/tenants.ts:
```diff
@@ -0,0 +1,55 @@
+    createTenant: builder.mutation<void, CreateTenantInput>({
+      query: (body) => ({
+        method: "POST",
+        url: "/api/ee/tenants",
```
don't we usually use singular nouns for API endpoints? `/api/ee/tenant` instead of `/api/ee/tenants` ?

iethree commented on frontend/src/metabase/common/components/MappingEditor/MappingEditor.tsx:
```diff
@@ -51,35 +66,39 @@ export interface MappingEditorProps {
+  if (entries.filter((e) => e.key === key).length > 1) {
+    return t`Attribute keys can't have the same name`;
+  }
+  if (entries.some((e) => e.key.startsWith("@"))) {
+    return t`Keys starting with "@" are reserved for system use`;
+  }
```
little bug: if any key errors, they all error

![Screenshot 2025-07-01 at 1 03 57 PM](https://github.com/user-attachments/assets/922d8ce9-a474-4569-8561-4dd6c4ce1f60)

npfitz commented on e2e/support/cypress_data.js:
```diff
@@ -41,10 +41,14 @@ export const SAMPLE_DB_TABLES = {
   ALL_USERS_GROUP: 1,
   ADMIN_GROUP: 2,
-  COLLECTION_GROUP: 3,
-  DATA_GROUP: 4,
-  READONLY_GROUP: 5,
-  NOSQL_GROUP: 6,
+  COLLECTION_GROUP: 4,
+  DATA_GROUP: 5,
+  READONLY_GROUP: 6,
+  NOSQL_GROUP: 7,
+};
```
Really, this doesn't even need to exist... the exact same information is available through `cypress_sample_instace_data.js`. Should we follow this up with a PR removing it?

npfitz commented on frontend/src/metabase/admin/people/containers/PeopleListingApp.tsx:
```diff
@@ -9,24 +10,33 @@ import { AdminPaneLayout } from "metabase/common/components/AdminPaneLayout";
 const PAGE_SIZE = 25;
 
-export function PeopleListingApp({ children }: { children: React.ReactNode }) {
+export function PeopleListingApp({
+  children,
+  external = false,
+}: {
```
I don't *love* it, but it works. I hate that we end up passing `external` down the component tree. But re-making everything felt silly too

npfitz commented on frontend/src/metabase/admin/permissions/constants/messages.ts:
```diff
@@ -6,6 +6,10 @@ export const UNABLE_TO_CHANGE_ADMIN_PERMISSIONS = t`Administrators always have t
 export const NATIVE_PERMISSION_REQUIRES_DATA_ACCESS = t`Groups with View data access set to "Blocked" can't create queries.`;
 // eslint-disable-next-line ttag/no-module-declaration -- see metabase#55045
 export const UNABLE_TO_CHANGE_LEGACY_PERMISSIONS = t`Change "No self-service (Deprecated)" View data access to enable custom Create queries permissions.`;
+// eslint-disable-next-line ttag/no-module-declaration -- see metabase#55045
+export const EXTERNAL_USERS_NO_ACCESS_COLLECTION = t`External Users can only access tenant collections`;
+// eslint-disable-next-line ttag/no-module-declaration -- see metabase#55045
+export const EXTERNAL_USERS_NO_ACCESS_DATABASE = t`External Users cannot manage database permissions`;
```
I actually implemented a fix for all of them but shoving them in a map with 
```
{
  get EXTERNAL_USER_MESSAGE() {
    return t`stuff...`;
  }
}
```

npfitz commented on frontend/src/metabase/admin/people/containers/PeopleListingApp.tsx:
```diff
@@ -36,66 +46,48 @@ export function PeopleListingApp({ children }: { children: React.ReactNode }) {
+                      <Button variant="filled">{t`Invite someone`}</Button>
+                    </Link>
+                  )}
+                  {!external && (
+                    <PLUGIN_TENANTS.EditUserStrategySettingsButton />
+                  )}
```
I had asked the oracle to reconsider. They will tell me more later haha

npfitz commented on enterprise/frontend/src/metabase-enterprise/tenants/containers/TenantsListingApp.tsx:
```diff
@@ -0,0 +1,46 @@
+  const [searchInputValue, setSearchInputValue] = useState("");
+  const [status, setStatus] = useState(ACTIVE_STATUS.active);
+
+  const { isLoading, error, data } = useListTenantsQuery({ status });
+  const tenants = useMemo(() => data?.data ?? [], [data]);
```
That's a good point. We probably do want to do something here. Will create a follow up issue 👍

npfitz commented on frontend/src/metabase/admin/people/components/ActiveStatusFilter.tsx:
```diff
@@ -0,0 +1,26 @@
+  status: ActiveStatus;
+  onStatusChange: (status: ActiveStatus) => void;
+}
+export const ActiveStatusFilter = ({
+  status,
+  onStatusChange,
+}: ActiveStatusFilterProps) => {
+  return (
+    <Radio.Group
+      value={status}
+      onChange={(val) => onStatusChange(ACTIVE_STATUS[val])}
+    >
+      <Group>
+        <Radio label={t`Active`} value={ACTIVE_STATUS.active} />
+        <Radio label={t`Deactivated`} value={ACTIVE_STATUS.deactivated} />
+      </Group>
+    </Radio.Group>
+  );
+};
```
Another good follow up task. I'll ask an adult haha

npfitz commented on enterprise/frontend/src/metabase-enterprise/tenants/components/TenantDisplayName/TenantDisplayName.tsx:
```diff
@@ -0,0 +1,19 @@
+    return <Loader />;
+  }
+
+  const currentTenant = tenants.data.find((tenants) => tenants.id === id);
```
Passing the tenant in wouldn't 100% be better. Just not sure how with plugins lol

npfitz committed:
whitespaces is hard

edpaget committed:
feat: add support for setting tenant attributes (#60096)

* feat: add support for setting tenant attributes

* fixup! PR feedback

* fix docstring
* fix put update name check to also check id
* return attributes in tenant-present

* fixup! mistake in migration

edpaget committed:
feat: merge tenant attributes into user attributes (#60153)

* feat: merge tenant attributes into user attributes

* fixup! use user attribute mapping

Co-Authored-By: Dan Sutton <dan@metabase.com>

* fixup! tenant login attributes docstring

Closes ADM-1098

* fixup! remove merging tenant/user attributes in api

* fixup! swap add-attributes order and add tests

---------

Co-authored-by: Dan Sutton <dan@metabase.com>

iethree committed:
allow tenant attributes CRUD (#60347)

wip merge user and tenant attributes

WIP use new api

Split Mapping Editor

types fixed

so much unit testing

handle null cases

update tests

update a couple more e2e tests

less silly test

make things nicer

invalidate user cache on tenant update

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/api/tenants.ts

edpaget committed:
feat: make tenanting attribute editable on FE (#61235)

Hide it behind the tenanting setting being turned on.

edpaget committed:
feat: tenant group api updates (#60866)

* feat: allow tenant groups to be created thru api

* feat: allow group filtering on internal/external

* feat: external users cannot be group managers

Add tests for other tenant group rules

* fix: validate external users is included

ADM-1203

edpaget committed:
fix: tenant groups can have subscription app perm (#61100)

iethree committed:
Tenant Groups UI (#60887)

* UI for external groups

* design updates

* all

* bug fixes and tests

# Conflicts:
#	e2e/test/scenarios/admin-2/tenants.cy.spec.js

* update test
